pkg/metrics/collector/sockstat.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2018 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type sockStatCollector struct {
+	inuse  *prometheus.Desc
+	tw     *prometheus.Desc
+	memory *prometheus.Desc
+}
+
+var pageSize = os.Getpagesize()
+
+func init() {
+	registerCollector("socket", NewSockStatCollector)
+}
+
+// NewSockStatCollector returns a new Collector exposing socket stats.
+func NewSockStatCollector() (Collector, error) {
+	return &sockStatCollector{
+		inuse: prometheus.NewDesc(
+			"num_inuse_sockets",
+			"Number of inuse sockets",
+			[]string{"protocol"}, nil,
+		),
+		tw: prometheus.NewDesc(
+			"num_tw_sockets",
+			"Number of sockets in time wait state",
+			nil, nil,
+		),
+		memory: prometheus.NewDesc(
+			"socket_memory",
+			"Amount of memory used by sockets",
+			nil, nil,
+		),
+	}, nil
+}
+
+func initSocketStats() map[string]uint64 {
+	stats := map[string]uint64{
+		"tcp_inuse": 0,
+		"udp_inuse": 0,
+		"tcp_tw":    0,
+		"mem":       0,
+	}
+	return stats
+}
+
+func getStatsName(protocol, name string) string {
+	// convert the name to merge data for a given stat
+	if name == "mem" {
+		return name
+	}
+
+	if protocol[len(protocol)-1] == '6' {
+		protocol = protocol[:len(protocol)-1]
+	}
+	return strings.ToLower(protocol) + "_" + name
+}
+
+func procSocketStats(sockStats map[string]map[string]string, result map[string]uint64) error {
+	for protocol, protocolStats := range sockStats {
+		for name, value := range protocolStats {
+			sname := getStatsName(protocol, name)
+			if old, ok := result[sname]; ok {
+				v, err := strconv.ParseUint(value, 10, 64)
+				if err != nil {
+					return fmt.Errorf("invalid value %s in sockstats: %s", value, err)
+				} else {
+					result[sname] = old + v
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func getSockStats(fileName string) (map[string]map[string]string, error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseSockStats(file)
+}
+
+func parseSockStats(r io.Reader) (map[string]map[string]string, error) {
+	sockStat := make(map[string]map[string]string)
+	scanner := bufio.NewScanner(r)
+
+	for scanner.Scan() {
+		line := strings.Split(scanner.Text(), " ")
+		// Remove trailing ':'.
```

Comment:
shouldn't we check that it has a trailing ':' before blindly chopping it off?

Reply:
Added more checks.